File: boj_codes/Python3/PyPy3/18111.py
# https://www.acmicpc.net/problem/18111
# 20201016

import sys

N, M, B = map(int, sys.stdin.readline().split())

def calc_blocks(blocks, height):
    cost = 0
    req_block = 0

    for i in range(N):
        for j in range(M):
            D = blocks[i][j] - height

            if D > 0:
                cost += 2 * D
                req_block -= D
            elif D < 0:
                cost -= D
                req_block -= D

    return cost, req_block

blocks = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

max_H = max(map(max, blocks))
min_H = min(map(min, blocks))
smallest_cost = 999999999
smallest_H = 0


for H in range(min_H, max_H + 1):
    cost, req_block = calc_blocks(blocks, H)

    if req_block > B:
        break

    elif cost <= smallest_cost:
            smallest_cost = cost
            smallest_H = H

print(smallest_cost, smallest_H)

File: boj_codes/C++17/18111.cpp
// https://www.acmicpc.net/problem/18111
// 20201019

#include <stdio.h>

void calc_blocks(int **blocks, int height, int N, int M, int &cost, int &req_block){
    int D;
    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            D = blocks[i][j] - height;

            if (D > 0){
                cost += 2 * D;
                req_block -= D;
            }
            else if (D < 0){
                cost -= D;
                req_block -= D;
            }
        }
    }

    return;
}

int main(){
    int N, M, B;
    int cost = 0, req_block = 0;
    int smallest_cost = __INT32_MAX__;
    int min_h = 256, max_h = 0;
    int cur_h, smallest_h;
    
    scanf("%d %d %d", &N, &M, &B);
    int** blocks = new int*[N];
    for(int i = 0; i < N; i++){
        blocks[i] = new int[M];
    } 

    for(int i = 0; i < N; i++){
        for(int j = 0; j < M; j++){
            scanf("%d", &blocks[i][j]);
            cur_h = blocks[i][j];
            if (cur_h < min_h)
                min_h = cur_h;
            if (cur_h > max_h)
                max_h = cur_h;
        }
    }

    for(int h = min_h; h <= max_h; h++){
        calc_blocks(blocks, h, N, M, cost, req_block);

        if (req_block > B){
            cost = 0;
            req_block = 0;
            break;
        }
        
        else if(cost <= smallest_cost){
            smallest_cost = cost;
            smallest_h = h;
        }

        cost = 0;
        req_block = 0;
    }
    
    printf("%d %d", smallest_cost, smallest_h);

    for(int i = 0; i < N; i++){
        delete [] blocks[i];
    } 
    delete [] blocks;  

    return 0;
}